tests/Authenticator.hs

```diff
@@ -0,0 +1,426 @@
+      -- NOTE: We do not perform user tests, instead assuming that the user always consents.
+      forM_ excludeCredentialDescriptorList $ \descriptor -> case authenticatorLookupCredential authenticator (M.pkcdId descriptor) of
+        Just (_, item) -> do
+          when (rpId == pkcsRpId item) (fail "InvalidStateError")
+        Nothing -> pure ()
+
+      -- 4. If requireResidentKey is true and the authenticator cannot store a
+      -- client-side discoverable public key credential source, return an error
+      -- code equivalent to "ConstraintError" and terminate the operation.
+      -- NOTE: We do not have to do this because the test authenticator supports
+      -- both methods of discoverability
+
+      -- 5. If requireUserVerification is true and the authenticator cannot
+      -- perform user verification, return an error code equivalent to
+      -- "ConstraintError" and terminate the operation.
+      -- NOTE: We do not do this because we fake user verification
+
+      -- 6. Collect an authorization gesture confirming user consent for creating
+      -- a new credential. The prompt for the authorization gesture is shown by
+      -- the authenticator if it has its own output capability, or by the user
+      -- agent otherwise. The prompt SHOULD display rpEntity.id, rpEntity.name,
+      -- userEntity.name and userEntity.displayName, if possible.
+      --   If requireUserVerification is true, the authorization gesture MUST
+      --   include user verification.
+      --   If requireUserPresence is true, the authorization gesture MUST include a
+      --   test of user presence.
+      --   If the user does not consent or if user verification fails, return an
+      --   error code equivalent to "NotAllowedError" and terminate the operation.
+      -- NOTE: We curently always succeed this step
+
+      -- 7. Once the authorization gesture has been completed and user consent
+      -- has been obtained, generate a new credential object:
+      -- 7.1. Let (publicKey, privateKey) be a new pair of cryptographic keys
+      -- using the combination of PublicKeyCredentialType and cryptographic
+      -- parameters represented by the first item in credTypesAndPubKeyAlgs that
+      -- is supported by this authenticator.
+      let Just param = find ((`elem` aSupportedAlgorithms) . M.pkcpAlg) credTypesAndPubKeyAlgs
+      (publicKey, privateKey) <- newKeyPair $ M.pkcpAlg param
+
+      -- 7.2. Let userHandle be userEntity.id.
+      let userHandle = M.pkcueId userEntity
+
+      -- 7.4 If requireResidentKey is true or the authenticator chooses to
+      -- create a client-side discoverable public key credential source:
+
+      -- 7.4.1
+      -- NOTE: We need to have a CredentialId before we can construct the
+      -- credentialSource.
+      -- NOTE: We always construct a clientside discoverable credential, as this
+      -- is allowed.
+      credentialId <- M.CredentialId <$> Random.getRandomBytes 16
```

Comment:
We should add at least a TODO here, because we're currently using two ways to generate randomness in the library:
- The library implements `Uniform CredentialId`, which comes from the `random` library, which the server part uses for randomness.
- The tests here use cryptonite's `MonadRandom`

By default we should be using a secure random source, but for tests we specifically want a deterministic random source

Reply:
Problem is that the authenticator uses MonadRandom to generate new keys. We could possibly generate new keys ourselves, bit I'm not sure on the complexity.